nexus/reconfigurator/planning/src/mgs_updates/mod.rs

```diff
@@ -476,6 +652,179 @@ fn try_make_update_sp(
+                        ArtifactKind::GIMLET_ROT_IMAGE_B,
+                        ArtifactKind::PSC_ROT_IMAGE_B,
+                        ArtifactKind::SWITCH_ROT_IMAGE_B,
+                    ];
+
+                    if slot_b_artifacts.contains(&a.id.kind) {
+                        return true;
+                    }
+                }
+            }
+
+            false
+        })
+        .collect();
+    if matching_artifacts.is_empty() {
+        warn!(
+            log,
+            "cannot configure RoT update for board (no matching artifact)";
+            baseboard_id,
+        );
+        return None;
+    }
+
+    if matching_artifacts.len() > 1 {
+        // This should be impossible unless we shipped a TUF repo with more
+        // than 1 artifact for the same board and slot. But it doesn't prevent
+        // us from picking one and proceeding. Make a note and proceed.
+        warn!(log, "found more than one matching artifact for RoT update");
+    }
+
+    let artifact = matching_artifacts[0];
+
+    // If the artifact's version matches what's deployed, then no update is
+    // needed.
+    if artifact.id.version == expected_active_version {
+        debug!(log, "no RoT update needed for board"; baseboard_id);
+        return None;
+    }
+
+    let expected_active_slot = ExpectedActiveRotSlot {
+        slot: active_slot,
+        version: expected_active_version,
+    };
+
+    // Begin configuring an update.
+    let inactive_caboose = match active_slot.toggled() {
+        RotSlot::A => CabooseWhich::RotSlotA,
+        RotSlot::B => CabooseWhich::RotSlotB,
+    };
+
+    let expected_inactive_version = match inventory
+        .caboose_for(inactive_caboose, baseboard_id)
+        .map(|c| c.caboose.version.parse::<ArtifactVersion>())
+        .transpose()
+    {
+        Ok(None) => ExpectedVersion::NoValidVersion,
+        Ok(Some(v)) => ExpectedVersion::Version(v),
+        Err(_) => {
+            warn!(
+                log,
+                "cannot configure RoT update for board \
+                 (found inactive slot contents but version was not valid)";
+                baseboard_id
+            );
+            return None;
+        }
+    };
+
+    Some(PendingMgsUpdate {
+        baseboard_id: baseboard_id.clone(),
+        sp_type: sp_info.sp_type,
+        slot_id: u32::from(sp_info.sp_slot),
+        details: PendingMgsUpdateDetails::Rot {
+            expected_active_slot,
+            expected_inactive_version,
+            expected_persistent_boot_preference: rot_state
+                .persistent_boot_preference,
+            expected_pending_persistent_boot_preference: rot_state
+                .pending_persistent_boot_preference,
+            expected_transient_boot_preference: rot_state
+                .transient_boot_preference,
+        },
+        artifact_hash: artifact.hash,
+        artifact_version: artifact.id.version.clone(),
+    })
+}
+
 #[cfg(test)]
 mod test {
```

Comment:
I think we could augment a lot of the unit tests here and add new ones to cover the new RoT logic.